### PR TITLE
i.sentinel.download: use EODAG as a unified API for downloading sentinel products

### PR DESCRIPTION
**Pull request by HamedElgizery**
This PR will reimplement i.sentinel.download making use of the i.eodag module to use the EODAG API as the main source of Sentinel products.

### CODE REVIEW

File: src/imagery/i.sentinel/i.sentinel.download/i.sentinel.download.py
Comment (lucadelu) [inline src/imagery/i.sentinel/i.sentinel.download/i.sentinel.download.py]: @HamedElgizery are you sure this is needed? is the config file already exists this is superfluous

**Comment by lucadelu** [comment]
It seems a big simplification, as soon it will be ready to test let us know.

File: src/imagery/i.sentinel/i.sentinel.download/i.sentinel.download.py
Comment (HamedElgizery) [reply src/imagery/i.sentinel/i.sentinel.download/i.sentinel.download.py]: I think it is important because Copernicus Data Space Ecosystem, which has almost all the products, requires credentials for searching. So it can be an indication for the user to fill in the credentials.
However, I think having the settings file option back is the best workaround, but I need to implement it in i.eodag first...

File: src/imagery/i.sentinel/i.sentinel.download/i.sentinel.download.py
Comment (griembauer) [inline src/imagery/i.sentinel/i.sentinel.download/i.sentinel.download.py]: In the case `"GCS": "earth_search_gcs"` I suppose google credentials are required to be configured via eodag? In the former version of `i.sentinel.download` this was not required because the GCS download URL was automatically composed by a given scene ID.
I think it is a great solution to use eodag as a one-stop-shop now, but if google credentials are required, this would also have to be indicated in the .html (or it is also possible to omit the Google cloud storage version entirely)

Comment (griembauer) [reply src/imagery/i.sentinel/i.sentinel.download/i.sentinel.download.py]: Sorry for annoying so early - I noticed you only just began with this draft PR ;) I am looking forward to this solution!

Comment (HamedElgizery) [reply src/imagery/i.sentinel/i.sentinel.download/i.sentinel.download.py]: Hello @griembauer :D

> (or it is also possible to omit the Google cloud storage version entirely)

I think that would be a reasonable choice since Google cloud storage only provides one Sentinel product in the EODAG context (S2_MSI_L1C) which is already provided by Copernicus.

**Comment by HamedElgizery** [comment]
@neteler @ninsbl 
I was looking for these two products **S3OL1RAC & S3OL1SAC** on EODAG but they don't exist on Copernicus. 
I did find the first one, S3OL1RAC, on two other providers (WEKEO & SARA), but for the second product, S3OL1SAC, I don't seem to find it anywhere. Do you have any suggestion about them and if they are necessary to have?

File: src/imagery/i.sentinel/i.sentinel.download/i.sentinel.download.py
Comment (veroandreo) [inline src/imagery/i.sentinel/i.sentinel.download/i.sentinel.download.py]: This was a pilot L2 product IIRC, @neteler do you think we can safely remove?

Comment (neteler) [reply src/imagery/i.sentinel/i.sentinel.download/i.sentinel.download.py]: Yes, IMHO no need to keep pilot products.

File: src/imagery/i.sentinel/i.sentinel.download/i.sentinel.download.py
Comment (ninsbl) [inline src/imagery/i.sentinel/i.sentinel.download/i.sentinel.download.py]: Instead of an if-in you can use the get method here...

File: src/imagery/i.sentinel/i.sentinel.download/i.sentinel.download.py
Comment (lucadelu) [reply src/imagery/i.sentinel/i.sentinel.download/i.sentinel.download.py]: Thanks @griembauer, it is correct to add to documentation since it is an important change

File: src/imagery/i.sentinel/i.sentinel.download/i.sentinel.download.py
Comment (HamedElgizery) [reply src/imagery/i.sentinel/i.sentinel.download/i.sentinel.download.py]: > @HamedElgizery are you sure this is needed? is the config file already exists this is superfluous

I found a way to have the settings option back, without needing to add anything to i.eodag.
I have put back the settings option, and removed the config file check in these two commits [d32374b](https://github.com/OSGeo/grass-addons/pull/1150/commits/d32374bdab064b750149090f650d00a77afbd762) [70f5c87](https://github.com/OSGeo/grass-addons/pull/1150/commits/70f5c87b52a4bc13c33fd63d998fa00ea3cc1c0a)
I am using the environment variables for that.

File: src/imagery/i.sentinel/i.sentinel.download/i.sentinel.download.py
Comment (veroandreo) [inline src/imagery/i.sentinel/i.sentinel.download/i.sentinel.download.py]: I guess this needs editing too if we are not using the uuid anymore?

File: src/imagery/i.sentinel/i.sentinel.download/i.sentinel.download.py
Comment (veroandreo) [inline src/imagery/i.sentinel/i.sentinel.download/i.sentinel.download.py]: GCS appears as deprecated below, shouldn't this be changed then?

File: src/imagery/i.sentinel/i.sentinel.download/i.sentinel.download.py
Comment (veroandreo) [inline src/imagery/i.sentinel/i.sentinel.download/i.sentinel.download.py]: S1 products do not have cloudcover, that's right because they are SAR data. But S3 is optic afaik, don't they have cloud cover property?? Just checking, haven't used S3.

File: src/imagery/i.sentinel/i.sentinel.download/i.sentinel.download.py
Comment (veroandreo) [inline src/imagery/i.sentinel/i.sentinel.download/i.sentinel.download.py]: Shouldn't this be a single message? Users will need to have i.eodag installed which requires eodag.

**Review by veroandreo — COMMENTED** [review]
I left a couple of comments. Would you mind updating the manual page too, so it's easier to test?

**Comment by HamedElgizery** [comment]
> I left a couple of comments. Would you mind updating the manual page too, so it's easier to test?

Thanks @veroandreo, I will update the manual ASAP.

File: src/imagery/i.sentinel/i.sentinel.download/i.sentinel.download.py
Comment (ninsbl) [inline src/imagery/i.sentinel/i.sentinel.download/i.sentinel.download.py]: Instead of duplicating the dict with values as keys, you could use a generator to get the dict key for an eodag product like:

```
product_to_find = "S3_SY_SYN"
key = next(option_key for option_key, eodag_product in PRODUCTTYPE_MAP.items() if eodag_product == product_to_find)
```

**Comment by ninsbl** [comment]
> Do you have any suggestion about them and if they are necessary to have?

S3 OL 1 RAC is an internal product. Perfectly fine to drop it... Not sure what the background of the SAC product has been. But I did not find documentation either. So fine to drop as well...

File: src/imagery/i.sentinel/i.sentinel.download/i.sentinel.download.py
Comment (ninsbl) [inline src/imagery/i.sentinel/i.sentinel.download/i.sentinel.download.py:247]: GRD data are now provided in SAFE format with internal COG, see: https://github.com/CS-SI/eodag/pull/1277

This illutrates how hardcoding product types limits `i.sentinel.download` compared to `i.eodag`.

You should add a note in the manual that i.eodag is used under the hood and offers more comprehensive download possibilities...

**Review by ninsbl — CHANGES_REQUESTED** [review]
A general question is how backwards-compatibility is handled or if we should put a lot of effort into it?

Some features will disappear in this PR (e.g. possibility to download from DHUS portals = sentinelsat).

Personally, I would give priority to basic tests in i.eodag...

File: src/imagery/i.sentinel/i.sentinel.download/i.sentinel.download.py
Comment (HamedElgizery) [reply src/imagery/i.sentinel/i.sentinel.download/i.sentinel.download.py]: After some testing, these are the Sentinel-3 products that have cloud coverage percentage:
* S3SY2SYN
* S3SY2VGP
* S3SY2VG1
* S3SY2V10
* S3SY2AOD

The rest do have the cloud coverage attribute but it is always 0.

File: src/imagery/i.sentinel/i.sentinel.download/i.sentinel.download.py
Comment (HamedElgizery) [reply src/imagery/i.sentinel/i.sentinel.download/i.sentinel.download.py:247]: > GRD data are now provided in SAFE format with internal COG, see: [CS-SI/eodag#1277](https://github.com/CS-SI/eodag/pull/1277)

Cool, shall I add it or wait till it is officially released? 

> This illutrates how hardcoding product types limits `i.sentinel.download` compared to `i.eodag`.

Agree I think in most cases it is almost always better to use i.eodag directly, for more flexibility. Except for pre-existing code I guess.

> You should add a note in the manual that i.eodag is used under the hood and offers more comprehensive download possibilities...
 
Will do that.

**Comment by HamedElgizery** [comment]
Thanks for the review, @ninsbl.

> Personally, I would give priority to basic tests in i.eodag...

I will be working on that in parallel with this PR now. 
This PR should be ready soon, as I think what is remaining for the most part is:
1. Updating the manual.
2. Emulate the -p flag to give a similar output as the previous implementation.

Is there any points that I might have not recognized?

File: src/imagery/i.sentinel/i.sentinel.download/i.sentinel.download.py
Comment (ninsbl) [reply src/imagery/i.sentinel/i.sentinel.download/i.sentinel.download.py:247]: > > GRD data are now provided in SAFE format with internal COG, see: [CS-SI/eodag#1277](https://github.com/CS-SI/eodag/pull/1277)
> 
> Cool, shall I add it or wait till it is officially released?

there is a pre-release out. So I guess it is OK to add it...

**Comment by ninsbl** [comment]
> I will be working on that in parallel with this PR now. This PR should be ready soon, as I think what is remaining for the most part is: (...)
> 2. Emulate the -p flag to give a similar output as the previous implementation.

For me it is no problem if the format of printed info changes a bit after switching to eodag... Not an issue I would invest a lot of time in. As mentioned, I think there will be breaking changes anyway...

File: src/imagery/i.sentinel/i.sentinel.download/i.sentinel.download.html
Comment (ninsbl) [inline src/imagery/i.sentinel/i.sentinel.download/i.sentinel.download.html]:
```suggestion
(e.g. if a product has the timeliness queryable set as <b>Null</b>,
```

File: src/imagery/i.sentinel/i.sentinel.download/i.sentinel.download.html
Comment (ninsbl) [inline src/imagery/i.sentinel/i.sentinel.download/i.sentinel.download.html]: I would mention it here...

File: src/imagery/i.sentinel/i.sentinel.download/i.sentinel.download.html
Comment (ninsbl) [inline src/imagery/i.sentinel/i.sentinel.download/i.sentinel.download.html]: Data on CSDE is "Immediately available data (IAD)". Ao no archive. Not sure about other providers. You can remove the entire section I guess, but I would mention the providers used somewhere...

File: src/imagery/i.sentinel/i.sentinel.download/i.sentinel.download.html
Comment (ninsbl) [inline src/imagery/i.sentinel/i.sentinel.download/i.sentinel.download.html:142]: File name?

File: src/imagery/i.sentinel/i.sentinel.download/i.sentinel.download.html
Comment (ninsbl) [inline src/imagery/i.sentinel/i.sentinel.download/i.sentinel.download.html:207]: There are still a couple of references in the examples and in the python script...

File: src/imagery/i.sentinel/i.sentinel.download/i.sentinel.download.html
Comment (ninsbl) [inline src/imagery/i.sentinel/i.sentinel.download/i.sentinel.download.html:208]: It would be great if you could structure examples with sub-captions, e.g. by Satellte (h3 for Sentinel-1, -2, and -3, printing providers, products queriables. ...).

File: src/imagery/i.sentinel/i.sentinel.download/i.sentinel.download.html
Comment (HamedElgizery) [reply src/imagery/i.sentinel/i.sentinel.download/i.sentinel.download.html]: > but I would mention the providers used somewhere...

CDSE is the only provider used in the script currently, this is what you meant right?

Comment (HamedElgizery) [reply src/imagery/i.sentinel/i.sentinel.download/i.sentinel.download.html]: > > but I would mention the providers used somewhere...
> 
> CDSE is the only provider used in the script currently, this is what you mean right?

**Comment by veroandreo** [comment]
There are unresolved conversations here, a merging conflict and then we need an approval to merge. Can you please check the unresolved conversations and merge conflict @HamedElgizery ? After that I'd merge and do a second round of review once all is in

**Comment by ninsbl** [comment]
As @veroandreo mentioned, backwards compatibility is not that a hard requirement for addons, But I think users should get a heads-up (GRASS-dev / GRASS-users ML) before breaking changes are submittet.

**Comment by veroandreo** [comment]
> As @veroandreo mentioned, backwards compatibility is not that a hard requirement for addons, But I think users should get a heads-up (GRASS-dev / GRASS-users ML) before breaking changes are submittet.

I agree we can do a  courtesy heads up, but that's not common practice for addons really. COAH and sentinelsat stopped working almost one year ago, so users of i.sentinel should already know, I suppose. 

I am moreover for merging this, update the manual with a notice about COAH and sentinelsat not working anymore, and sending an email afterwards.

**Comment by ninsbl** [comment]
> I am moreover for merging this, update the manual with a notice about COAH and sentinelsat not working anymore, and sending an email afterwards.

No objections from my side...

**Review by ninsbl — APPROVED** [review]

**Comment by wenzeslaus** [comment]
Please wait for https://github.com/OSGeo/grass/pull/4205 to be merged, then restart the failing job.

**Pull request merged**
